atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java

```diff
@@ -652,6 +633,97 @@ public String toString() {
+
+                List<ColumnOrSuperColumn> cells = cellsForRow.getValue();
+
+                if (!cells.isEmpty()) {
+                    ByteBuffer row = cellsForRow.getKey();
+                    result.getOrDefault(row, new ArrayList<>()).addAll(cells);
+                    query.add(new KeyPredicate()
+                            .setKey(row)
+                            .setPredicate(getSlicePredicate(cells)));
+                }
+
+            }
+        }
+
+        Map<Cell, Value> ret = Maps.newHashMapWithExpectedSize(rows.size());
+        new ValueExtractor(metricsManager, ret)
+                .extractResults(result, startTs, ColumnSelection.all());
+        return ret;
+    }
+
+    //todo(Sudiksha): rename | refactor
+    private Map<ByteBuffer, List<ColumnOrSuperColumn>> fetchCellsForKeyPredicates(final InetSocketAddress host,
+            final TableReference tableRef,
+            List<KeyPredicate> query,
+            final long startTs) throws Exception {
+        return clientPool.runWithRetryOnHost(
+                host,
+                new FunctionCheckedException<CassandraClient, Map<ByteBuffer, List<ColumnOrSuperColumn>>, Exception>() {
+                    @Override
+                    public Map<ByteBuffer, List<ColumnOrSuperColumn>> apply(CassandraClient client) throws Exception {
+
+                        //todo(Sudiksha): double check logging
+                        if (log.isTraceEnabled()) {
+                            log.trace("Requesting {} cells from {} {}starting at timestamp {} on {}",
+                                    SafeArg.of("cells", query.size()),
+                                    LoggingArgs.tableRef(tableRef),
+                                    SafeArg.of("timestampClause", ""),
+                                    SafeArg.of("startTs", startTs),
+                                    SafeArg.of("host", CassandraLogHelper.host(host)));
+                        }
+
+                        Map<ByteBuffer, List<List<ColumnOrSuperColumn>>> results = wrappingQueryRunner.multiget_multislice(
+                                "getRows", client, tableRef, query, readConsistency);
+                        Map<ByteBuffer, List<ColumnOrSuperColumn>> aggregatedResults = Maps.transformValues(results,
+                                lists -> Lists.newArrayList(Iterables.concat(lists)));
+
+                        return aggregatedResults;
+                    }
+
+                    @Override
+                    public String toString() {
+                        //todo(Sudiksha): double check
+                        return "multiget_multislice(" + host + ", "
+                                + query.size() + " cells" + ")";
+                    }
+
+                }
+        );
+    }
+
+    //todo(Sudiksha): refactor | names
+    private static SlicePredicate getSlicePredicate(List<ColumnOrSuperColumn> columns) {
```

Comment:
at some point, you want to comment on the optimisation is that you're doing, it's easy to get lost in `CassandraKeyValueServiceImpl`. If you're able to encode that in your method names, even better, but might make sense to just comment/javadoc.